state/remote/s3.go

```diff
@@ -0,0 +1,119 @@
+package remote
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/service/s3"
+)
+
+func s3Factory(conf map[string]string) (Client, error) {
+	bucketName, ok := conf["bucket"]
+	if !ok {
+		return nil, fmt.Errorf("missing 'bucket' configuration")
+	}
+
+	keyName, ok := conf["key"]
+	if !ok {
+		return nil, fmt.Errorf("missing 'key' configuration")
+	}
+
+	regionName, ok := conf["region"]
+	if !ok {
+		regionName = os.Getenv("AWS_DEFAULT_REGION")
+		if regionName == "" {
+			return nil, fmt.Errorf("missing 'region' configuration or AWS_DEFAULT_REGION environment variable")
+		}
+	}
+
+	accessKeyId := conf["access_key"]
+	secretAccessKey := conf["secret_key"]
+
+	credentials := aws.DetectCreds(accessKeyId, secretAccessKey, "")
```

Comment:
One of the possible return values here is a provider that will, on request, try to hit the IAM role key provisioning endpoint to obtain a temporary key if Terraform is being run on a machine in EC2. Thus as far as I could tell it isn't possible to determine if the credentials object will work until the first request is made and it attempts to reach out to the provisioning endpoint.

I allowed for this case because I thought instance IAM keys would be pretty useful for granting only a specific set of build/deploy worker machines read/write access to the S3 bucket, without having to distribute user-level credentials.

However, I'd agree that the resulting user experience is a little odd if you happen to run without credentials on a non-EC2 machine. (It fails with a big error that talks about being unable to reach a URL on 169.254.169.254, which would be quite opaque to someone who doesn't know how EC2 meta-data/user-data work behind the scenes.)